README.md
# AnyEvent::Finger ![static](https://github.com/uperl/AnyEvent-Finger/workflows/static/badge.svg) ![linux](https://github.com/uperl/AnyEvent-Finger/workflows/linux/badge.svg)

Simple asynchronous finger client and server

# SYNOPSIS

client:

```perl
use AnyEvent::Finger qw( finger_client );

finger_client 'localhost', 'username', sub {
  my($lines) = @_;
  print "[response]\n";
  print join "\n", @$lines;
};
```

server:

```perl
use AnyEvent::Finger qw( finger_server );

my %users = (
  grimlock => 'ME GRIMLOCK HAVE ACCOUNT ON THIS MACHINE',
  optimus  => 'Freedom is the right of all sentient beings.',
);

finger_server sub {
  my $tx = shift; # isa AnyEvent::Finger::Transaction
  if($tx->req->listing_request)
  {
    # respond if remote requests list of users
    $tx->res->say('users: ', keys %users);
  }
  else
  {
    # respond if user exists
    if(defined $users{$tx->req->username})
    {
      $tx->res->say($users{$request});
    }
    # respond if user does not exist
    else
    {
      $tx->res->say('no such user');
    }
  }
  # required! done generating the reply,
  # close the connection with the client.
  $tx->res->done;
};
```

# DESCRIPTION

This distribution provides an asynchronous finger server and
client which can be used by any event loop supported by
[AnyEvent](https://metacpan.org/pod/AnyEvent).  This specific module provides a simple procedural
interface to client and server classes also in this distribution.

# FUNCTIONS

## finger\_client

```
finger_client( $server, $request, $callback, [ \%options ] )
```

Send a finger request to the given server.  The callback will
be called when the response is complete.  The options hash may
be passed in as the optional forth argument to override any
default options (See [AnyEvent::Finger::Client](https://metacpan.org/pod/AnyEvent::Finger::Client) for details).

## finger\_server

```perl
my $server = finger_server $callback, \%options;
```

Start listening to finger callbacks and call the given callback
for each request.  See [AnyEvent::Finger::Server](https://metacpan.org/pod/AnyEvent::Finger::Server) for details
on the options and the callback.

# CAVEATS

Finger is an oldish protocol and almost nobody uses it anymore.

Most finger clients do not have a way to configure an alternate port.
Binding to the default port 79 on Unix usually requires root.  Running
[AnyEvent::Finger::Server](https://metacpan.org/pod/AnyEvent::Finger::Server) as root is not recommended.

Under Linux you can use `iptables` to forward requests to port 79 to
an unprivileged port.  I was able to use this incantation to forward port 79
to port 8079:

```
# iptables -t nat -A PREROUTING -p tcp --dport 79 -j REDIRECT --to-port 8079
# iptables -t nat -I OUTPUT -p tcp -d 127.0.0.1 --dport 79 -j REDIRECT --to-port 8079
```

The first rule is sufficient for external clients, the second rule was required
for clients connecting via the loopback interface (localhost).

# SEE ALSO

- [RFC1288](http://tools.ietf.org/html/rfc1288),
- [AnyEvent::Finger::Client](https://metacpan.org/pod/AnyEvent::Finger::Client),
- [AnyEvent::Finger::Server](https://metacpan.org/pod/AnyEvent::Finger::Server)
- [AnyEvent](https://metacpan.org/pod/AnyEvent)

    Generic non-blocking event loop used by [AnyEvent::Finger](https://metacpan.org/pod/AnyEvent::Finger)

- [Net::Finger](https://metacpan.org/pod/Net::Finger)

    Blocking implementation of a finger client

- [Net::Finger::Server](https://metacpan.org/pod/Net::Finger::Server)

    Blocking implementation of a finger server

- [PlugAuth::Plugin::Finger](https://metacpan.org/pod/PlugAuth::Plugin::Finger)

    PlugAuth plugin that allows querying a PlugAuth server
    via the finger protocol.  Uses this module for its
    implementation.

# AUTHOR

Graham Ollis <plicease@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2012-2022 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.14      2022-09-25 13:25:52 -0600
  - Removed overload on AnyEvent::Finger::Transaction
    which was unintentional and worse incomplete (gh#6, gh#8)

0.12      2021-04-28 13:43:07 -0600
  - Fixed bug in overload in some of the classes (gh#3)

0.11      2020-02-05 14:20:27 -0700
  - Diagnostic release

0.10      2015-03-18 02:16:24 -0400
  - Documentation improvements

0.09      2014-04-08 14:38:42 -0400
  - Support for Perl 5.8

0.08      2013-10-06 01:15:35 -0400
  - documentation tweaks

0.07      2013-03-20 18:24:30 America/New_York
  - fixup meta resources

0.06      2013-03-20 17:48:44 America/New_York
  - documentation updates

0.05      2013-02-09 18:46:53 America/New_York
  - documentation tweaks

0.04      2013-02-09 17:00:20 America/New_York
  - minor inheritence tweak

0.03      2013-02-09 09:22:11 America/New_York
  - pass $server object into on_bind first argument

0.02      2013-02-09 09:13:27 America/New_York
  - create on_bind callback on server for when port # is not known in advance
  - don't use Condvar#recv from AnyEvent::Finger::Server as it may not be
    supported by the AnyEvent implementation

0.01      2013-01-22 10:50:03 America/New_York
  - initial version


lib/AnyEvent/Finger.pm
package AnyEvent::Finger;

use strict;
use warnings;
use base qw( Exporter );

our @EXPORT_OK = qw( finger_client finger_server );

# ABSTRACT: Simple asynchronous finger client and server
# VERSION

=head1 SYNOPSIS

client:

 use AnyEvent::Finger qw( finger_client );
 
 finger_client 'localhost', 'username', sub {
   my($lines) = @_;
   print "[response]\n";
   print join "\n", @$lines;
 };

server:

 use AnyEvent::Finger qw( finger_server );
 
 my %users = (
   grimlock => 'ME GRIMLOCK HAVE ACCOUNT ON THIS MACHINE',
   optimus  => 'Freedom is the right of all sentient beings.',
 );
 
 finger_server sub {
   my $tx = shift; # isa AnyEvent::Finger::Transaction
   if($tx->req->listing_request)
   {
     # respond if remote requests list of users
     $tx->res->say('users: ', keys %users);
   }
   else
   {
     # respond if user exists
     if(defined $users{$tx->req->username})
     {
       $tx->res->say($users{$request});
     }
     # respond if user does not exist
     else
     {
       $tx->res->say('no such user');
     }
   }
   # required! done generating the reply,
   # close the connection with the client.
   $tx->res->done;
 };

=head1 DESCRIPTION

This distribution provides an asynchronous finger server and
client which can be used by any event loop supported by
L<AnyEvent>.  This specific module provides a simple procedural
interface to client and server classes also in this distribution.

=head1 FUNCTIONS

=head2 finger_client

 finger_client( $server, $request, $callback, [ \%options ] )

Send a finger request to the given server.  The callback will
be called when the response is complete.  The options hash may
be passed in as the optional forth argument to override any
default options (See L<AnyEvent::Finger::Client> for details).

=cut

sub finger_client ($$$;$)
{
  my($hostname) = shift;
  require AnyEvent::Finger::Client;
  AnyEvent::Finger::Client
    ->new( hostname => $hostname )
    ->finger(@_);
  ();
}

=head2 finger_server

 my $server = finger_server $callback, \%options;

Start listening to finger callbacks and call the given callback
for each request.  See L<AnyEvent::Finger::Server> for details
on the options and the callback.

=cut

# keep the server object in scope so that
# we don't unbind from the port.  If you
# don't want this, then use the OO interface
# for ::Server instead.
my $keep = [];

sub finger_server ($;$)
{
  require AnyEvent::Finger::Server;
  my $server = AnyEvent::Finger::Server
    ->new
    ->start(@_);
  push @$keep, $server;
  $server;
}

1;

=head1 CAVEATS

Finger is an oldish protocol and almost nobody uses it anymore.

Most finger clients do not have a way to configure an alternate port.
Binding to the default port 79 on Unix usually requires root.  Running
L<AnyEvent::Finger::Server> as root is not recommended.

Under Linux you can use C<iptables> to forward requests to port 79 to
an unprivileged port.  I was able to use this incantation to forward port 79
to port 8079:

 # iptables -t nat -A PREROUTING -p tcp --dport 79 -j REDIRECT --to-port 8079
 # iptables -t nat -I OUTPUT -p tcp -d 127.0.0.1 --dport 79 -j REDIRECT --to-port 8079

The first rule is sufficient for external clients, the second rule was required
for clients connecting via the loopback interface (localhost).

=head1 SEE ALSO

=over 4

=item

L<RFC1288|http://tools.ietf.org/html/rfc1288>,

=item

L<AnyEvent::Finger::Client>,

=item

L<AnyEvent::Finger::Server>

=item

L<AnyEvent>

Generic non-blocking event loop used by L<AnyEvent::Finger>

=item

L<Net::Finger>

Blocking implementation of a finger client

=item

L<Net::Finger::Server>

Blocking implementation of a finger server

=item

L<PlugAuth::Plugin::Finger>

PlugAuth plugin that allows querying a PlugAuth server
via the finger protocol.  Uses this module for its
implementation.

=back

=cut
